ext: allow builders to override support for linkcode references

jdknight opened the pull request:
### Feature or Bugfix

- Feature

### Purpose

The original `sphinx.ext.linkcode` extension will only include references into a doctree for `html` builders. Custom builders who would like to handle linkcode references do not have a graceful way to enable this feature.

This commit aims to allow custom builders to override when these references are included. If a builder defines a `supported_linkcode` attribute, this extension will use the value of this attribute to consider other builders/formats when conditionally adding references.

### Relates
- https://github.com/sphinx-contrib/confluencebuilder/issues/421

tk0miya reviewed (commented):
+1: Reasonable. It's much better to describe this behavior in the document of linkcode extension.

AA-Turner commented on sphinx/ext/linkcode.py:
```diff
@@ -24,6 +24,8 @@ def doctree_read(app: Sphinx, doctree: Node) -> None:
         raise LinkcodeError(
             "Function `linkcode_resolve` is not given in conf.py")
 
+    node_only_expr = getattr(app.builder, 'supported_linkcode', 'html')
```
```suggestion
    # By default, the linkcode extension will only inject references
    # for a `html` builder. If a builder wishes to support managing
    # references generated by linkcode as well, it can define the 
    # `supported_linkcode` attribute.
    node_only_expr = getattr(app.builder, 'supported_linkcode', 'html')
```

jdknight committed:
[ext-linkcode] allow builders to flag support for linkcode references

The original `sphinx.ext.linkcode` extension will only include
references into a doctree for `html` builders. Custom builders who
would like to handle linkcode references do not have a graceful way to
enable this feature.

This commit aims to allow custom builders to override when these
references are included. If a builder defines a `supported_linkcode`
attribute, this extension will use the value of this attribute to
consider other builders/formats when conditionally adding references.

Signed-off-by: James Knight <james.d.knight@live.com>

jdknight committed:
[doc] add extdev documentation for interfacing with the linkcode ext

Adding a document under Sphinx's "extension development" chapter which
provides information about integrating third-party builds with recently
added features to the built-in linkcode extension.

Signed-off-by: James Knight <james.d.knight@live.com>
